### PR TITLE
I18 make this work with error codes

### LINKED ISSUE #18: Make this work with error codes.
Think of how to make it work.
```C#
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;

namespace Pryvicy
{
    public static class OperationResult
    {
        public sealed class SuccessResult<T> : IOperationResult<T>
        {
            public T Value { get; private set; }
            public SuccessResult(T value)
            {
                Value = value;
            }

            public IOperationResult<TOut> Match<TOut>(
                Func<T, IOperationResult<TOut>> bindSuccess,
                Func<string, IOperationResult<TOut>> bindError,
                Func<Exception, IOperationResult<TOut>> bindFailure)
            {
                return bindSuccess(Value);
            }

            public async Task<IOperationResult<TOut>> MatchAsync<TOut>(
                Func<T, Task<IOperationResult<TOut>>> bindSuccess,
                Func<string, Task<IOperationResult<TOut>>> bindError,
                Func<Exception, Task<IOperationResult<TOut>>> bindFailure)
            {
                return await bindSuccess(Value);
            }
        }

        public sealed class ErrorResult<T> : IOperationResult<T>
        {
            public string Message { get; private set; }
            public ErrorResult(string message)
            {
                Message = message;
            }

            public IOperationResult<TOut> Match<TOut>(
                Func<T, IOperationResult<TOut>> bindSuccess,
                Func<string, IOperationResult<TOut>> bindError,
                Func<Exception, IOperationResult<TOut>> bindFailure)
            {
                return bindError(Message);
            }

            public async Task<IOperationResult<TOut>> MatchAsync<TOut>(
                Func<T, Task<IOperationResult<TOut>>> bindSuccess,
                Func<string, Task<IOperationResult<TOut>>> bindError,
                Func<Exception, Task<IOperationResult<TOut>>> bindFailure)
            {
                return await bindError(Message);
            }
        }

        public sealed class FailureResult<T> : IOperationResult<T>
        {
            public Exception Exception { get; private set; }
            public FailureResult(Exception exception)
            {
                Exception = exception;
            }

            public IOperationResult<TOut> Match<TOut>(
                Func<T, IOperationResult<TOut>> bindSuccess,
                Func<string, IOperationResult<TOut>> bindError,
                Func<Exception, IOperationResult<TOut>> bindFailure)
            {
                return bindFailure(Exception);
            }

            public async Task<IOperationResult<TOut>> MatchAsync<TOut>(
                Func<T, Task<IOperationResult<TOut>>> bindSuccess,
                Func<string, Task<IOperationResult<TOut>>> bindError,
                Func<Exception, Task<IOperationResult<TOut>>> bindFailure)
            {
                return await bindFailure(Exception);
            }
        }

        public static IOperationResult<T> Success<T>(T value)
        {
            return new SuccessResult<T>(value);
        }

        public static IOperationResult<T> Error<T>(string message)
        {
            return new ErrorResult<T>(message);
        }

        public static IOperationResult<T> Failure<T>(Exception exception)
        {
            return new FailureResult<T>(exception);
        }

        public static IOperationResult<T> ToResult<T>(this T source)
        {
            return Success(source);
        }

        public static IOperationResult<TOut> Bind<TIn, TOut>(this IOperationResult<TIn> source, Func<TIn, IOperationResult<TOut>> bind)
        {
            return source.Match(
                value => bind(value),
                error => Error<TOut>(error),
                exception => Failure<TOut>(exception)
            );
        }

        public static IOperationResult<T> IfError<T>(this IOperationResult<T> source, Func<IOperationResult<T>> returnResult)
        {
            return source.Match(
                _ => source,
                error => returnResult(),
                failure => returnResult()
            );
        }

        public static async Task<IOperationResult<T>> IfErrorAsync<T>(this Task<IOperationResult<T>> source, Func<IOperationResult<T>> returnResult)
        {
            return (await source).IfError(returnResult);
        }

        public static IOperationResult<T> BindError<T>(this IOperationResult<T> source, Func<string, IOperationResult<T>> bind, Func<Exception, string> getMessage)
        {
            return source.Match(
                _ => source,
                error => bind(error),
                failure => bind(getMessage(failure))
            );
        }

        public static async Task<IOperationResult<T>> BindErrorAsync<T>(this Task<IOperationResult<T>> source, Func<string, IOperationResult<T>> bind, Func<Exception, string> getMessage)
        {
            return (await source).BindError(bind, getMessage);
        }

        public static IOperationResult<T> UseError<T>(this IOperationResult<T> source, Action<string> use, Func<Exception, string> getMessage)
        {
            return source.Match(
                _ => source,
                error =>
                {
                    use(error);
                    return source;
                    },
                failure =>
                {
                    use(getMessage(failure));
                    return source;
                }
            );
        }

        public static async Task<IOperationResult<T>> UseErrorAsync<T>(this IOperationResult<T> source, Func<string, Task> use, Func<Exception, string> getMessage)
        {
            return await source.MatchAsync(
                _ => Task.FromResult(source),
                async error =>
                {
                    await use(error);
                    return source;
                },
                async failure =>
                {
                    await use(getMessage(failure));
                    return source;
                }
            );
        }

        public static async Task<IOperationResult<T>> UseErrorAsync<T>(this Task<IOperationResult<T>> source, Func<string, Task> use, Func<Exception, string> getMessage)
        {
            return await (await source).UseErrorAsync(use, getMessage);
        }

        public static async Task<IOperationResult<T>> UseErrorAsync<T>(this Task<IOperationResult<T>> source, Action<string> use, Func<Exception, string> getMessage)
        {
            return (await source).UseError(use, getMessage);
        }

        public static IOperationResult<T> Use<T>(this IOperationResult<T> source, Action<T> sideEffect)
        {
            var success = source as SuccessResult<T>;
            if (success == null)
                return source;

            sideEffect(success.Value);

            return source;
        }

        public static async Task<IOperationResult<TOut>> BindAsync<TIn, TOut>(this IOperationResult<TIn> source, Func<TIn, Task<IOperationResult<TOut>>> bindAsync)
        {
            return await source.MatchAsync(
                async value => await bindAsync(value),
                error => Task.FromResult(Error<TOut>(error)),
                exception => Task.FromResult(Failure<TOut>(exception))
            );
        }

        public static async Task<IOperationResult<TOut>> BindAsync<TIn, TOut>(this Task<IOperationResult<TIn>> source, Func<TIn, Task<IOperationResult<TOut>>> bindAsync)
        {
            return await (await source).BindAsync(bindAsync);
        }

        public static async Task<IOperationResult<TOut>> BindAsync<TIn, TOut>(this Task<IOperationResult<TIn>> source, Func<TIn, IOperationResult<TOut>> bind)
        {
            return (await source).Bind(bind);
        }

        public static IOperationResult<T> Try<T>(Func<T> func)
        {
            try
            {
                return func().ToResult();
            }
            catch (Exception ex)
            {
                return Failure<T>(ex);
            }
        }

        public static async Task<IOperationResult<T>> TryAsync<T>(Func<Task<T>> func)
        {
            try
            {
                return (await func()).ToResult();
            }
            catch (Exception ex)
            {
                return Failure<T>(ex);
            }
        }

        public static IOperationResult<T[]> All<T>(this IEnumerable<Func<IOperationResult<T>>> source)
        {
            return source.Aggregate(new List<T>().ToResult(), (a, f) => a.Bind(aList => f().Bind(newRes =>
            {
                aList.Add(newRes);
                return aList.ToResult();
            }))).Bind(x => x.ToArray().ToResult());
        }
    }
}
```

### PR DESCRIPTION
**Pull request by TheSonOfDeimos**
Closes #18

### CODE REVIEW

File: src/Fls.Results/OperationResult.cs [hunk 1]
```diff
@@ -122,5 +141,44 @@ public static IOperationResult<T> Failure<T>(Exception exception)
         {
             return (await source).Bind(bind);
         }
+
+        public static async Task<IOperationResult<T>> BindErrorAsync<T>(this Task<IOperationResult<T>> source, Func<string, IOperationResult<T>> bind, Func<Exception, string> getMessage)
+        {
+            return (await source).BindError(bind, getMessage);
+        }
+        
+        public static async Task<IOperationResult<T>> BindErrorAsync<T>(this IOperationResult<T> source, Func<string, Task<IOperationResult<T>>> bind, Func<Exception, string> getMessage)
```
Comment (vkamiansky) [inline src/Fls.Results/OperationResult.cs:146]: These functions here look like special cases of  `public static async Task<IOperationResult<T>> BindErrorAsync<T>(this IOperationResult<T> source, Func<int?, string, Task<IOperationResult<T>>> bind, Func<Exception, string> getMessage, int? exceptionCode = null)`. Can't we make use of them?

**Review by vkamiansky — CHANGES_REQUESTED** [review]
Suggested review of `Bind` and `BindAsync` for default code overload.

File: src/Fls.Results/OperationResult.cs (hunk 1 above)
Comment (TheSonOfDeimos) [reply src/Fls.Results/OperationResult.cs:146]: Of course, we can use 
```C#
public static async Task<IOperationResult<T>> BindErrorAsync<T>(this IOperationResult<T> source, Func<int?, string, Task<IOperationResult<T>>> bind, Func<Exception, string> getMessage, int? exceptionCode = null)
```
but such signature makes us to define addtional empty parametr in delegate, like 
```C#
BindErrorAsync(
                (_, message) =>
                {
                    return IOperationResult<T>();
                },

                exception =>
                {
                    return default(string);
                }
            );
```

But with overloading, there is no need in fictitious parameter like here
```C#
BindErrorAsync(
                message =>
                {
                    return IOperationResult<T>();
                },

                exception =>
                {
                    return default(string);
                }
            );
```

Comment (TheSonOfDeimos) [reply src/Fls.Results/OperationResult.cs:150]: вп

Comment (TheSonOfDeimos) [reply src/Fls.Results/OperationResult.cs:146]: Also, if we have to support old `Bind` and `BindAsync` signatures, why don't we support old signatures of delegates?

Comment (vkamiansky) [reply src/Fls.Results/OperationResult.cs:146]: No worries. All I mean is that this method can be implemented as follows:
```c#
public static IOperationResult<T> BindError<T>(this IOperationResult<T> source, 
    Func<string, IOperationResult<T>> bind, 
    Func<Exception, string> getMessage)
    {
        return source.BindError((_, error) => bind(error), getMessage);
    }
```

Arguably, the same applies to other methods with 'no error code' signatures.

**Review by vkamiansky — CHANGES_REQUESTED** [review]
Just need some simplification.

File: src/Fls.Results/OperationResult.cs (hunk 1 above)
Comment (TheSonOfDeimos) [reply src/Fls.Results/OperationResult.cs:146]: Oh, you're right, that looks better

**Review by vkamiansky — CHANGES_REQUESTED** [review]
Almost there. Just walk through the .cs files with Alt+Shift+F to fix the formatting.

**Review by vkamiansky — APPROVED** [review]
Ready to merge.

**Pull request merged**